database/migrations/000032_provider_class.up.sql

```diff
@@ -0,0 +1,35 @@
+-- Copyright 2024 Stacklok, Inc
+--
+-- Licensed under the Apache License, Version 2.0 (the "License");
+-- you may not use this file except in compliance with the License.
+-- You may obtain a copy of the License at
+--
+--      http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+
+CREATE TYPE provider_class AS enum ('github', 'gh-app');
+
+ALTER TABLE providers
+  -- Fills existing rows with 'github'
+  ADD COLUMN class provider_class DEFAULT 'github' NOT NULL;
+
+-- providers becomes a star schema, where the provider is the central table,
+-- and different classes of providers have different provider_* tables that
+-- join with the providers table to get the credential parameters.
+
+CREATE TABLE provider_app_installations (
```

Comment:
To recap this table, is this where we add the unclaimed installations? How will the claim concept work with this?

Reply:
Makes sense to me, thanks for the recap!

Reply:
It sounds like there's a second flow where we're encoding some additional state in the callback URL.  I'm going to add that here much like it exists in `session_store`, but without needing a second table.

Reply:
For the `state` in the callback URL we can reuse the existing `session_store` table. The state will be used to lookup the provider and project. Once we do the lookup, we have no use for the state and should not store it here.

Reply:
Fair enough, I was going to clear the nonce at the same time as I set the `provider_id`, so both wouldn't be set at once.  But we can have a separate `session_store` row.

What is the `provider` parameter for the `session_store` in the OAuth flow case? I'm assuming we'd also ignore `owner_filter`.

Reply:
With the flow we're talking about, we wouldn't need an explicit `CreateProvider` RPC, since the provider create would be implicit in the conclusion of the OAuth flow.  (We may need to be able to name the providers automatically, but that seems feasible at the moment.)

Reply:
I'm going to re-use `session_store` for now for the session state, but I'm going to include a GitHub user ID in this table to allow indicating that a new project should be created for the specified user when they sign up.